Project goal for Contracts: primitive ownership assertions (#700)
rust-lang/rust-project-goals: src/2026/contracts-ownership-assertions.md

| Point of contact | @dawidl022                                   |
| Status           | Proposed                                                                         |
| Tracking issue   | #149213     |
| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |
| Needs | Funding |

## Summary

This project aims to implement the `owned` and `alloc_block` separation logic
assertions in the Rust compiler, as described in [MCP
#942](https://github.com/rust-lang/compiler-team/issues/942), to enable the
formal specification and verification of unsafe memory manipulation. The effort
is split into two main phases:

- Implementation: Building the core compiler intrinsics and higher-level wrappers, alongside integrating them with Miri for dynamic runtime testing.

- Documentation: Creating extensive reference materials and real-world tutorials to teach developers how to apply these separation logic concepts to verify libraries using unsafe Rust, e.g. the Rust standard library.

> **Needs funding:** This goal needs funding to proceed.

Reviewer: Do you want to add a section describing the funding you need?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..10cddc1a8f441ba04e4c387f01f823c36a84579c)*

Author: I'd rather keep the specifics out of the project goal itself. I have not seen other project goals expand on funding more than this one statement.

## Motivation

### The status quo

Safe Rust code depends on lower-level unsafe Rust libraries. The correctness of
these low-level libraries directly impacts the memory safety guarantees of safe
Rust code. By giving developers and formal methods specialists the language
necessary to express the correctness of unsafe Rust code, we advance a step
closer to closing the gap between the widely praised hypothetical safety of Rust
as a language, to tangible proofs of such being the case.

> 

Reviewer: I feel some context is needed about the contracts feature
```suggestion

There is currently an [experiment](https://github.com/rust-lang/rust/issues/128044) inside rustc to add special annotations that describe pre/post-conditions of functions. While built to be extensible, today this can only represent rather simple invariants. Unsafe code typically needs to reason with a more powerful logic, for example about memory ownership.

```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..10cddc1a8f441ba04e4c387f01f823c36a84579c)*

The first step in building the Rust specification language for use in
[contracts](../2025h1/std-contracts.md) is introducing ownership primitives, as
described in detail in [MCP
#942](https://github.com/rust-lang/compiler-team/issues/942). These ownership
assertions leverage decades of formal verification research on separation
logic, and have been implemented in practical systems recently ([Fulminate](https://dl.acm.org/doi/10.1145/3704879)), offering

> a promising direction for the verification landscape in Rust.

Reviewer: You're asserting things as necessities that aren't; this is better phrased as a proposal, e.g. something like:
```suggestion
We want this specification language to be able to specify
[the whole of the standard library](../2025h1/std-contracts.md).
For this purpose, we propose to introduce ownership primitives, as
described in detail in [MCP #942](https://github.com/rust-lang/compiler-team/issues/942).
These primitives are based on the standard and well-understood theory of separation logic,
and modeled after [Fulminate](https://dl.acm.org/doi/10.1145/3704879),
which is a practical, runtime-checkable, and non-expert-legible take on writing
separation logic predicates, which looks to be a great fit for Rust.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..10cddc1a8f441ba04e4c387f01f823c36a84579c)*

### Work items over the next year

#### Implementation

| Task        | Owner(s) | Notes |
| ----------- | -------- | ----- |
| Low-level ownership assertion intrinsics available for use in contracts. | @dawidl022  |       |
| Higher-level wrapper assertions for typical use cases, including alignment, arrays etc. | @dawidl022 | |
| Prototype integration with Miri - runtime ownership checks should work with the prototype Miri runtime ownership tracker as [implemented by Johannes Hostert](https://github.com/JoJoDeveloping/miri/tree/fulminate).|@dawidl022 | |
| Full Miri runtime ownership support | |

#### Documentation

| Task        | Owner(s) | Notes |
| ----------- | -------- | ----- |
| Reference level documentation: explaining the semantics of the ownership primitives, along with basic examples. | @dawidl022 | |
| A tutorial: selecting one of the challenges in https://model-checking.github.io/verify-rust-std/challenges.html and walking the developer through what it means to write the specifications, and testing them using the Miri runtime. | @dawidl022 | |

## Team asks

| Team       | Support level | Notes                                   |

> | ---------- | ------------- | --------------------------------------- |

Reviewer: I think you could add: from t-compiler you'll need reviews, and maybe from the miri team you'll need design and implementation work? Ideally from t-compiler you'd have @celinval agreeing to review your PRs since afaik she's the one who knows the contract infrastructure best.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..10cddc1a8f441ba04e4c387f01f823c36a84579c)*

# Contracts: primitive ownership assertions

| Metadata         |                                                                                  |
| :--------------- | -------------------------------------------------------------------------------- |
| What and why     | Implement [MCP #942](https://github.com/rust-lang/compiler-team/issues/942): experiment with Fulminate-like ownership primitives in contracts                |
| Point of contact | @dawidl022                                   |
| Status           | Proposed                                                                         |
| Tracking issue   |      |
| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |

> | Needs | Funding |

Reviewer 2: This field (and the "Needs funding" paragraph below) is outdated; please add a `## Funding` section according to the [template](https://github.com/rust-lang/rust-project-goals/blob/main/src/TEMPLATE.md).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..6dc77207ba575e1f5a8201b5816b0cb1f05be87e)*

Author: Thanks, updated.

standard library to be given contract specifications, bringing us further
towards the goal of [specifying and proving correct the entire Rust standard
library](../2024h2/std-verification.md). This addition specifically targets code
using `unsafe` Rust, which is known to have subtle correctness arguments, and
serves as the low level building blocks for many fundamental data structures in
the standard library.

Since many developers may be unfamiliar with Rust contracts and separation
logic, we will also put together documentation and a tutorial as part of this
goal. The benefits of doing this are two-fold: 1. we can directly evaluate the applicability of `owned` and `alloc_block` in real-world code (stdlib) and 2. it walks developers through non-trivial examples, allowing them to better understand the concepts and be able to apply them well to their own code, and be able to flag shortcomings in the current approach. This would be similar in spirit to the Nomicon (https://doc.rust-lang.org/nomicon/) and "Learn Rust With Entirely Too Many Linked Lists" (https://rust-unofficial.github.io/too-many-lists/index.html).

### Work items over the next year

> #### Implementation
> 
> | Task        | Owner(s) | Notes |
> | ----------- | -------- | ----- |
> | Low-level ownership assertion intrinsics available for use in contracts. | @dawidl022  |       |
> | Higher-level wrapper assertions for typical use cases, including alignment, arrays etc. | @dawidl022 | |
> | Prototype integration with Miri - runtime ownership checks should work with the prototype Miri runtime ownership tracker as [implemented by Johannes Hostert](https://github.com/JoJoDeveloping/miri/tree/fulminate).|@dawidl022 | |
> | Full Miri runtime ownership support | |
> 
> #### Documentation
> 
> | Task        | Owner(s) | Notes |
> | ----------- | -------- | ----- |
> | Reference level documentation: explaining the semantics of the ownership primitives, along with basic examples. | @dawidl022 | |
> | A tutorial: selecting one of the challenges in https://model-checking.github.io/verify-rust-std/challenges.html and walking the developer through what it means to write the specifications, and testing them using the Miri runtime. | @dawidl022 | |

Reviewer 2: For the full Miri runtime ownership support task, please add a `## Help wanted` section referencing that task. See [this goal](https://rust-lang.github.io/rust-project-goals/2026/libtest-json.html) for an example.

```suggestion
| Task        | Owner(s) | Notes |
| ----------- | -------- | ----- |
| Implement low-level ownership assertion intrinsics | @dawidl022  | |
| Implement higher-level wrapper assertions | @dawidl022 | Should cover typical use cases, including alignment, arrays, etc. |
| Implement a prototype integration with Miri | @dawidl022 | Runtime ownership checks should work with the prototype Miri runtime ownership tracker as [implemented by Johannes Hostert](https://github.com/JoJoDeveloping/miri/tree/fulminate). |
| Implement full Miri runtime ownership support | | |
| Write Reference-level documentation | @dawidl022 | Should explain the semantics of the ownership primitives, along with basic examples. |
| Write a tutorial | @dawidl022 | Select one of the challenges in https://model-checking.github.io/verify-rust-std/challenges.html, and walk the developer through what it means to write specifications and test them using the Miri runtime. |
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..6dc77207ba575e1f5a8201b5816b0cb1f05be87e)*

Author: Thanks, I've instead added myself as the owner of the task, as it would be within the scope of contributing to this goal.

## Team asks

| Team       | Support level | Notes                                   |
| ---------- | ------------- | --------------------------------------- |
| [compiler] | Small         | Code reviews                            |

> | [miri]     | Medium        | Guidance on integration                 |

Reviewer 2: Who will champion from miri?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/9d4289b4135587bc2626057a3db38a4d0a4b67a1..6dc77207ba575e1f5a8201b5816b0cb1f05be87e)*

Reviewer 2: This is being discussed: [#t-lang/contracts > Project Goal @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/544080-t-lang.2Fcontracts/topic/Project.20Goal/near/608376625), see also [#project-goals/2026-workshop > ✔ Contracts: primitive ownership assertions](https://rust-lang.zulipchat.com/#narrow/channel/546987-project-goals.2F2026-workshop/topic/.E2.9C.94.20Contracts.3A.20primitive.20ownership.20assertions/with/608999272)

Author: @RalfJung has agreed to champion from miri. Does this need to be recorded in the project goal markdown?

Reviewer 2: I added him

# Contracts: primitive ownership assertions

| Metadata         |                                                                                  |
| :--------------- | -------------------------------------------------------------------------------- |
| What and why     | Implement [MCP #942](https://github.com/rust-lang/compiler-team/issues/942): experiment with Fulminate-like ownership primitives in contracts                |
| Point of contact | @dawidl022                                   |
| Status           | Proposed                                                                         |
| Tracking issue   |      |

> | Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |

Reviewer 2: ```suggestion
| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |
| [miri] champion | @RalfJung |
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/a1c8150751053184075a765a6af67e8f08d0673c..86596ec3fa2a3d616d43fdf7f3458d492bf26e56)*

# Contracts: primitive ownership assertions

| Metadata         |                                                                                  |
| :--------------- | -------------------------------------------------------------------------------- |
| What and why     | Implement [MCP #942](https://github.com/rust-lang/compiler-team/issues/942): experiment with Fulminate-like ownership primitives in contracts                |
| Point of contact | @dawidl022                                   |
| Status           | Proposed                                                                         |
| Tracking issue   |      |
| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |

> | [miri] champion | @RalfJung |

Reviewer: ```suggestion
| [miri] champion | @RalfJung |
| [lang] champion | @Nadrieril |
```
This is a language experiment so should probably have a lang champion. Maybe this needs a lang ask too, something like "guidance on the design"?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-project-goals/700/a1c8150751053184075a765a6af67e8f08d0673c..7f29d0da8e42b26cf5817732a6cfda1b53247b1a)*

Reviewer 2: How do you feel about the ask size? Small, Medium?

Reviewer: Feels small, but then again a small ask is supposed to not need a champion so I feel like I'm breaking the rules ^^

Reviewer 2: Wait, I thought this was a compiler MCP?

Reviewer: Well, there was a bit of a mixup; that MCP was probably not the best vector for that particular request. Contracts are an experimental language feature with me as lang champion. The difficulty is that there are a few actors interested in contracts, and there wasn't a clear way to ask for their opinion on a design; we now have t-lang/contracts on Zulip for that.

Reviewer 3: If there can be more than one champion per team I'm happy to also wear my t-lang advisor hat for this. ;)

Author: Thanks @Nadrieril, I've added you as the lang champion and added a lang ask.
